### PR TITLE
feat(query): query cache table function

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Support `result_scan` table function for query cache.
- [x] implementation
- [x] tests
- [x] docs

**NOTE**: Since `result_scan` need dynamic table schema, so it is created in a different way than the normal table functions.

Closes #issue

### CODE REVIEW

File: src/query/storages/result_cache/src/table_function/table.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    #[async_trait::unboxed_simple]
+    async fn generate(&mut self) -> Result<Option<DataBlock>> {
+        if self.finish {
+            return Ok(None);
+        }
+
+        self.finish = true;
+        if self.ctx.get_settings().get_enable_query_result_cache()? {
+            if let Some(ref blocks) = self.blocks {
+                if self.ctx.get_settings().get_enable_query_result_cache()? {
```
Comment (sundy-li) [inline src/query/storages/result_cache/src/table_function/table.rs:178]: duplicated

Comment (ariesdevil) [reply src/query/storages/result_cache/src/table_function/table.rs]: done

File: src/query/storages/result_cache/src/table_function/table.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            if let Some(meta_key) = meta_key {
+                let kv_store = UserApiProvider::instance().get_meta_store_client();
+                let cache_reader = ResultCacheReader::create_with_meta_key(meta_key, kv_store);
+
+                let blocks = cache_reader.try_read_cached_result().await?;
+                return match blocks {
+                    Some(blocks) => Ok(Some(DataBlock::concat(&blocks)?)),
+                    None => Ok(None),
+                };
+            }
```
Comment (RinChanNOWWW) [inline src/query/storages/result_cache/src/table_function/table.rs:191]: It seems that this codes are unreachable.

File: src/query/service/src/interpreters/interpreter_select_v2.rs [hunk 3]
```diff
@@ -185,22 +200,43 @@ impl Interpreter for SelectInterpreterV2 {
...
+            // 2) select * from result_scan(last_query_id()); --> returns result same as line 1
+            // 3) insert into t1 values(2);
+            // 4) select * from t1; --> result changed since we insert new data.
+            // 5) select * from result_scan(last_query_id()); --> result same as line 2 cause cache
+            // If we read cache for 5, we will see it returns same result as 1 and 2 cause the
+            // generated result_cache_key are same for this statement, so here we fetch the previous
+            // meta_key through related query_id and using this meta_key to read cache.
+            let meta_key = if let Some(t) = self.result_scan_table() {
+                let arg_query_id = parse_result_scan_args(&t.table_args().unwrap())?;
+                self.ctx.get_result_cache_key(&arg_query_id)
```
Comment (RinChanNOWWW) [inline src/query/service/src/interpreters/interpreter_select_v2.rs:216]: After creating `ResultScanTable`, we can use the original pipeline (don't need to create cache reader again).

**Comment by ariesdevil** [comment]
Many thx to @RinChanNOWWW suggestions:)

File: src/query/catalog/src/plan/datasource/datasource_info/result_scan.rs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+use std::sync::Arc;
+
+use common_expression::TableSchema;
+use common_meta_app::schema::TableInfo;
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, Debug)]
+pub struct ResultScanTableInfo {
+    pub table_info: TableInfo,
+    pub query_id: String,
+    pub blocks_row_data: Vec<u8>,
```
Comment (RinChanNOWWW) [inline src/query/catalog/src/plan/datasource/datasource_info/result_scan.rs:24]: block_raw_data?

File: src/query/storages/result_cache/src/table_function/table.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        _plan: &DataSourcePlan,
+        pipeline: &mut Pipeline,
+    ) -> Result<()> {
+        if self.blocks_raw_data.is_empty() {
+            pipeline.add_source(EmptySource::create, 1)?;
+        } else {
+            let mut reader = Cursor::new(self.blocks_raw_data.clone());
+            let meta = read_metadata(&mut reader)?;
+            let arrow_schema = infer_schema(&meta)?;
+            let table_schema = TableSchema::from(&arrow_schema);
```
Comment (RinChanNOWWW) [inline src/query/storages/result_cache/src/table_function/table.rs:133]: Don't need to infer schema again. The schema is already in `self.table_info`.

**Review by sundy-li — APPROVED** [review]

**Comment by haydenflinner** [comment]
```
  2023-06-09T15:44:51.012409Z ERROR databend_query::servers::mysql::writers::query_result_writer: OnQuery Error: Code: 2004, Text = error: 
  --> SQL:1:27
  |
1 | SELECT * FROM RESULT_SCAN(LAST_QUERY_ID(-2))
  |                           ^^^^^^^^^^^^^^^^^ Expect Int64, but got Int8

.

<Backtrace disabled by default. Please use RUST_BACKTRACE=1 to enable> 
    at src/query/service/src/servers/mysql/writers/query_result_writer.rs:309
```


```
select version();
+---------------------------------------------------------------------------------------------------------------------+
| version()                                                                                                           |
+---------------------------------------------------------------------------------------------------------------------+
| 8.0.26-v1.1.54-nightly-21716fa733225e87377dd30af03999b3510cabc9(rust-1.70.0-nightly-2023-06-08T03:29:20.086461983Z) |
+---------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

**Pull request merged**